starboard/shared/libiamf/iamf_audio_decoder.cc

```diff
@@ -0,0 +1,382 @@
+      return "IAMF_ERR_INVALID_PACKET";
+    case IAMF_ERR_INVALID_STATE:
+      return "IAMF_ERR_INVALID_STATE";
+    case IAMF_ERR_UNIMPLEMENTED:
+      return "IAMF_ERR_UNIMPLEMENTED";
+    case IAMF_ERR_ALLOC_FAIL:
+      return "IAMF_ERR_ALLOC_FAIL";
+    default:
+      return "Unknown IAMF error code " + std::to_string(code);
+  }
+}
+}  // namespace
+
+IamfAudioDecoder::IamfAudioDecoder(const AudioStreamInfo& audio_stream_info,
+                                   bool prefer_binaural_audio)
+    : audio_stream_info_(audio_stream_info),
+      prefer_binarual_audio_(prefer_binaural_audio) {
+  decoder_ = IAMF_decoder_open();
+  if (!decoder_) {
+    SB_DLOG(ERROR) << "Error creating libiamf decoder";
+  }
+}
+
+IamfAudioDecoder::~IamfAudioDecoder() {
+  TeardownCodec();
+}
+
+bool IamfAudioDecoder::is_valid() const {
+  return decoder_ != NULL;
+}
+
+void IamfAudioDecoder::Initialize(const OutputCB& output_cb,
+                                  const ErrorCB& error_cb) {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(output_cb);
+  SB_DCHECK(!output_cb_);
+  SB_DCHECK(error_cb);
+  SB_DCHECK(!error_cb_);
+
+  output_cb_ = output_cb;
+  error_cb_ = error_cb;
+}
+
+void IamfAudioDecoder::Decode(const InputBuffers& input_buffers,
+                              const ConsumedCB& consumed_cb) {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(!input_buffers.empty());
+  SB_DCHECK(pending_audio_buffers_.empty());
+  SB_DCHECK(output_cb_);
+
+  if (stream_ended_) {
+    SB_DLOG(ERROR) << "Decode() is called after WriteEndOfStream() is called.";
+    return;
+  }
+
+  if (input_buffers.size() > kMinimumBuffersToDecode) {
+    std::copy(std::begin(input_buffers), std::end(input_buffers),
+              std::back_inserter(pending_audio_buffers_));
+    consumed_cb_ = consumed_cb;
+    DecodePendingBuffers();
+  } else {
+    for (const auto& input_buffer : input_buffers) {
+      if (!DecodeInternal(input_buffer)) {
+        return;
+      }
+    }
+    Schedule(consumed_cb);
+  }
+}
+
+void IamfAudioDecoder::DecodePendingBuffers() {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(!pending_audio_buffers_.empty());
+  SB_DCHECK(consumed_cb_);
+
+  for (int i = 0; i < kMinimumBuffersToDecode; ++i) {
+    if (!DecodeInternal(pending_audio_buffers_.front())) {
+      return;
+    }
+    pending_audio_buffers_.pop_front();
+    if (pending_audio_buffers_.empty()) {
+      Schedule(consumed_cb_);
+      consumed_cb_ = nullptr;
+      if (stream_ended_) {
+        Schedule(std::bind(&IamfAudioDecoder::WriteEndOfStream, this));
+        stream_ended_ = false;
+      }
+      return;
+    }
+  }
+
+  SB_DCHECK(!pending_audio_buffers_.empty());
+  Schedule(std::bind(&IamfAudioDecoder::DecodePendingBuffers, this));
+}
+
+bool IamfAudioDecoder::DecodeInternal(
+    const scoped_refptr<InputBuffer>& input_buffer) {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(input_buffer);
+  SB_DCHECK(input_buffer->size() > 0);
```

Comment:
We shouldn't assume the size of the input buffer.  An input buffer of size 0 is probably invalid, but we should trigger an error instead of DCHECK (crash in Devel).